### ISSUE TITLE
write() should not return the error from write_func()

### ISSUE THREAD

**Issue by gopakumarce** [kornelski/rust-security-framework#138]
First of all thanks a lot for this lib, helps a lot writing stuff for apple ecosystem.

As far as I understand from apple documentation, SSLWrite() will call SSLWriteFunc(), but if there is any kind of error returned by SSLWriteFunc(), SSLWrite() itself will return an error of its own, ie there is no responsibility the coder has to "save" the error returned by SSLWriteFunc() and then check that after calling SSLWrite() etc.. In fact, doing that might be erroneous - consider this example. 

1. We called SSLWrite(), SSLWriteFunc() might have returned errSslWouldBlock, but SSLWrite() itself succeeded because SSL lib is gonna buffer up the data we provide (and encrypt it and call SSLWriteFunc() on the encrypted data) - so the library has now "cached" an error errSslWouldBlock in conn.err
2. Later again we call SSLWrite(), this time just as an example say there was some error and the API returned errSSLPeerCertExpired
3. Now write() will go and check conn.err and will find errSslWouldBlock and that is what the caller to write() gets as an error, which is incorrect

So in summary, IMO there is no need to cache conn.err and return it - unless there was a reason for it, which ill be curious to know

**Comment by kornelski** [comment]
It would be helpful if you could write a test that demonstrates the problem.

**Comment by gopakumarce** [comment]
Sure, I will attempt a test case. But does my line of reasoning sound logical ? Leaving aside all the Rust wrappers, the matter under discussion here is simply "should SSLWriteFunc() error be saved, should callers of SSLWrite() check for the saved error ?" and to me it sounds that the obvious answer is no, because SSLWrite() itself can/will check for the error return on SSLWriteFunc() and translate/return that appropriately as needed. What I was curious to know is whether there is a specific reason you have done so in the  rust code or it was just something that was not necessarily for handling anything particular

**Comment by kornelski** [comment]
Yes, that makes sense. The [docs](https://developer.apple.com/documentation/security/sslwritefunc) say the result is passed on.
